packages/@glimmer/runtime/lib/compiled/opcodes/expressions.ts

```diff
@@ -395,3 +395,49 @@ APPEND_OPCODES.add(Op.GreaterEqual, (vm) => {
     })
   );
 });
+
+APPEND_OPCODES.add(Op.And, (vm) => {
+  let { positional } = check(vm.stack.popJs(), CheckArguments).capture();
+  let stack = vm.stack;
+
+  stack.pushJs(
+    createComputeRef(() => {
+      let values = reifyPositional(positional);
+
+      // this is handlebars definition of truthiness
+      for (let i = 0, len = values.length; i < len; i++) {
+        if (isTruthy(values[i]) === false) {
+          return values[i];
+        }
+      }
+      return values[values.length - 1];
+    })
+  );
+});
+
+APPEND_OPCODES.add(Op.Or, (vm) => {
+  let { positional } = check(vm.stack.popJs(), CheckArguments).capture();
+  let stack = vm.stack;
+
+  stack.pushJs(
+    createComputeRef(() => {
+      let values = reifyPositional(positional);
+
+      // this is handlebars definition of truthiness
+      for (let i = 0, len = values.length; i < len; i++) {
+        if (isTruthy(values[i]) === true) {
+          return values[i];
+        }
+      }
+      return values[values.length - 1];
+    })
+  );
+});
+
+function isTruthy(result: unknown) {
+  if (Array.isArray(result)) {
+    return result.length !== 0;
```

Comment:
I have some concerns about this. If this isn't a done deal I'd like to raise them. Where should I do that?

Reply:
I guess for now I can just say them here. Suppose someone had the following:

```hbs
{{or someMaybeEmptyArray otherArray}}
```

If we're trying to do TS types for this we can't just convert to `someMaybeEmptyArray || otherArray`. As long as `someMaybeEmptyArray` isn't `null` or `undefined` TS will see it as being truthy and will just return the type as `someMaybeEmptyArray` however, the correct type would actually be `someMaybeEmptyArray | otherArray`. In the case where someone wants to handle empty arrays, I think it would more correctly done as:

```hbs
{{if someMaybeEmptyArray.length someMaybeEmptyArray otherArray}}
```

We could then correctly type that as `someMaybeEmptyArray | otherArray`.